pkg/torcx/profile.go

```diff
@@ -215,3 +216,82 @@ func ListProfiles(profileDirs []string) (map[string]string, error) {
 
 	return profiles, nil
 }
+
+func mergeProfiles(applyCfg *ApplyConfig) (Images, error) {
+	var upperImages, lowerImages Images
+
+	if applyCfg == nil {
+		return Images{}, errors.New("missing apply configuration")
+	}
+
+	// Get lower profile (vendor)
+	if applyCfg.LowerProfile != "" {
+		opp, err := os.Open(VendorProfile)
+		if err != nil {
+			return Images{}, errors.Wrapf(err, "opening vendor profile", VendorProfile)
+		}
+		defer opp.Close()
+		lowerImages, err = readProfileReader(bufio.NewReader(opp))
+		if err != nil {
+			return Images{}, errors.Wrapf(err, "reading vendor profile", VendorProfile)
+		}
+	}
+
+	// Get upper profile (user)
+	if applyCfg.UpperProfile != "" {
+		localProfiles, err := ListProfiles(applyCfg.ProfileDirs())
+		if err != nil {
+			return Images{}, errors.Wrap(err, "profiles listing failed")
+		}
+		lowerProfilePath, ok := localProfiles[applyCfg.UpperProfile]
+		if !ok {
+			return Images{}, errors.Errorf("profile %q not found", applyCfg.UpperProfile)
+		}
+		opp, err := os.Open(lowerProfilePath)
+		if err != nil {
+			return Images{}, errors.Wrapf(err, "opening user profile", lowerProfilePath)
+		}
+		defer opp.Close()
+		upperImages, err = readProfileReader(bufio.NewReader(opp))
+		if err != nil {
+			return Images{}, errors.Wrapf(err, "reading user profile", VendorProfile)
+		}
+	}
+
+	return mergeImages(lowerImages, upperImages), nil
+
+}
+
+func mergeImages(lower Images, upper Images) Images {
+	// TODO(lucab): perhaps trade-off time/memory here with a linked-hashmap
+	merged := Images{Images: make([]Image, 0, len(lower.Images)+len(upper.Images))}
+	lowerImages := make(map[string]bool, len(lower.Images))
+	upperImages := make(map[string]bool, len(upper.Images))
+
+	// Compute the set of images to keep
+	for _, image := range lower.Images {
+		if image.Reference != "" {
+			lowerImages[image.Name] = true
+		}
+	}
+	for _, image := range upper.Images {
+		delete(lowerImages, image.Name)
+		if image.Reference != "" {
+			upperImages[image.Name] = true
+		}
+	}
+
+	// Merge in order
```

Comment:
Because I would need a map which records insertion order to later iterate on it (see linked-hashmap comment at the top). The rationale is that we want to merge upper *on top* of lower, overriding vendor entries but otherwise just appending user ones without any re-ordering. Any specific concern on this? I guess you'd like such a comment in the docstring ;)